Normalize block values before passing to TF provider

VenelinMartinov opened the pull request:
part of https://github.com/pulumi/pulumi-terraform-bridge/issues/1785

This change adds a normalisation step for collections when recovering cty values to pass to terraform. This ensures we represent them similarly to terraform.

In practice this means that all block collections need to be passed to TF providers as an empty collection instead of nil. This should get rid of quite a few subtle discrepancies in the data we pass to the TF provider code. These sometimes result in panics since we pass unexpected nils.

This gets rid of all known input discrepancies discovered so far through cross-testing.

The full rules for what is a block are [here](https://github.com/hashicorp/terraform-plugin-sdk/blob/1f499688ebd9420768f501d4ed622a51b2135ced/helper/schema/core_schema.go#L60). It is essentially properties with schema: typeList or typeSet with a Resource Elem.

fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1970
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1915
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1964
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1767 
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1762 

TODO: [DONE] remove the MaxItemsOne default hacks introduced in https://github.com/pulumi/pulumi-terraform-bridge/pull/1725 (opened https://github.com/pulumi/pulumi-terraform-bridge/pull/2049)

t0yv0 commented on pkg/tests/cross-tests/input_cross_test.go:
```diff
@@ -219,3 +219,27 @@ func TestInputsEmptyConfigModeAttrSet(t *testing.T) {
+				},
+			},
+		},
+		Config: tftypes.NewValue(tftypes.Object{}, map[string]tftypes.Value{}),
```
This is very strange, it's not that the set is empty, it's that it's not defined at all. So it should be treated as missing (which per Optional: true should be fine). Any more logs as to what's going on differently here? Certain these are not test-level issues ?

t0yv0 replied:
BTW. I'm not sure this `tftypes.NewValue(tftypes.Object{}, map[string]tftypes.Value{}),` is right at all. We need tftypes.Object to match 

```
Schema: map[string]*schema.Schema{
				"f0": {
					Optional: true,
					Type:     schema.TypeSet,
					Elem: &schema.Resource{
						Schema: map[string]*schema.Schema{
							"x": {Optional: true, Type: schema.TypeString},
						},
					},
				},
			},
```

Even if it is empty. That is the normal situation.

I think the framework will let you write the value as `map[string]any{}` and it will infer the correct tftypes.Object if you do so.

VenelinMartinov replied:
I think it's what https://github.com/pulumi/pulumi-terraform-bridge/issues/1964 is hitting.

pulumi program:

```yaml
backend:
    url: file://./data
name: project
resources:
    example:
        properties: {}
        type: crossprovider:index:TestRes
runtime: yaml
```

TF program

```terraform
resource "crossprovider_testres" "example" {
}
```

VenelinMartinov replied:
The program seem fine to me, so it looks like a bridge issue

t0yv0 replied:
Fair, the program looks fine. What is the bridge issue. Can you attach logs etc.

VenelinMartinov replied:
I haven't dug into it yet, will let you know once I have something more.

VenelinMartinov replied:
I think it happens here: https://github.com/pulumi/pulumi-terraform-bridge/blob/7aa41da657b39cb1791d936bd53d6fe06dc3d52c/pkg/tfshim/sdk-v2/cty.go#L232

t0yv0 reviewed (approved):
Good test case to add, continues to be surprising.

VenelinMartinov commented:
Pretty sure this is happening here: https://github.com/pulumi/pulumi-terraform-bridge/blob/7aa41da657b39cb1791d936bd53d6fe06dc3d52c/pkg/tfshim/sdk-v2/cty.go#L230


<details>

```
panic: here!

goroutine 201 [running]:
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2.recoverCtyValueOfObjectType({{0x104741db8, 0x14000782f20}}, 0x14000b64ea0)
	/Users/vvm/code/pulumi-terraform-bridge/pkg/tfshim/sdk-v2/cty.go:232 +0x2b0
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2.recoverCtyValue({{0x104741db8, 0x14000782f20}}, {0x104336680, 0x14000b64ea0})
	/Users/vvm/code/pulumi-terraform-bridge/pkg/tfshim/sdk-v2/cty.go:122 +0x264
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2.recoverAndCoerceCtyValueWithSchema({0x104730060, 0x14000ecbf20}, {0x104336680, 0x14000b64ea0})
	/Users/vvm/code/pulumi-terraform-bridge/pkg/tfshim/sdk-v2/cty.go:94 +0x40
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2.(*planResourceChangeImpl).Diff(0x14000996470, {0x1047419e8, 0x14000703a40}, {0x10392c6f9, 0x15}, {0x0?, 0x0?}, {0x10471c9a0?, 0x14000f29ef0?}, {0x0?, ...})
	/Users/vvm/code/pulumi-terraform-bridge/pkg/tfshim/sdk-v2/provider2.go:147 +0x10c
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2.(*providerWithPlanResourceChangeDispatch).Diff(0x1400077eb70, {0x1047419e8, 0x14000703a40}, {0x10392c6f9, 0x15}, {0x0, 0x0}, {0x10471c9a0, 0x14000f29ef0}, {0x0?, ...})
	/Users/vvm/code/pulumi-terraform-bridge/pkg/tfshim/sdk-v2/provider2.go:688 +0xb0
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge.(*Provider).Create(0x14000bb9988, {0x1047419e8?, 0x14000703380?}, 0x14000777450)
	/Users/vvm/code/pulumi-terraform-bridge/pkg/tfbridge/provider.go:1099 +0x5c0
github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Create_Handler({0x1046c0e40, 0x14000bb9988}, {0x1047419e8, 0x14000703380}, 0x14000a79080, 0x0)
	/Users/vvm/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.115.2/proto/go/provider_grpc.pb.go:586 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000aa2a00, {0x1047419e8, 0x140007032f0}, {0x1047553e0, 0x1400117a180}, 0x140003acd80, 0x14000ac47b0, 0x105986f00, 0x0)
	/Users/vvm/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1369 +0xb58
google.golang.org/grpc.(*Server).handleStream(0x14000aa2a00, {0x1047553e0, 0x1400117a180}, 0x140003acd80)
	/Users/vvm/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1780 +0xb20
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	/Users/vvm/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1019 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 191
	/Users/vvm/go/pkg/mod/google.golang.org/grpc@v1.63.2/server.go:1030 +0x13c
exit status 2
FAIL	github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tests/cross-tests	1.771s
```

</details>

t0yv0 commented:
attrs[attrName] = cty.NullVal(attrType) is a correct representation AFAIK, digging in.

VenelinMartinov commented:
In the cross tests this is flagged as a diff:

```
TF value cty.ObjectVal(map[string]cty.Value{"f0":cty.SetValEmpty(cty.Object(map[string]cty.Type{"x":cty.String})), "id":cty.UnknownVal(cty.String)})
    input_check.go:42: PU value cty.ObjectVal(map[string]cty.Value{"f0":cty.NullVal(cty.Set(cty.Object(map[string]cty.Type{"x":cty.String}))), "id":cty.NullVal(cty.String)})
```

t0yv0 commented:
Yeah this is plausible.

```
    input_check.go:41: TF value cty.ObjectVal(map[string]cty.Value{"f0":cty.SetValEmpty(cty.Object(map[string]cty.Type{"x":cty.String})), "id":cty.NullVal(cty.String)})
    input_check.go:42: PU value cty.ObjectVal(map[string]cty.Value{"f0":cty.NullVal(cty.Set(cty.Object(map[string]cty.Type{"x":cty.String}))), "id":cty.NullVal(cty.String)})

```

Perhaps this comes back to the fact that null values aren't representable in TF for set-nested blocks. Therefore it's an automatic empty set aka `[]`.

VenelinMartinov commented:
Yeah and I think this is were it gets really messy with SchemaConfigMode, since they are representable if ConfigModeAttr is set.

VenelinMartinov commented:
I ran into some trouble with `proposed_new.go` - not obvious how to convert the element type between hcty and cty

t0yv0 commented:
True - this is just saying the fix need to accommodate both for both cases:

- your code is testing a Set-nested block
- if is also possible to have a Set({foo: string}) attribute in TF, we should test that we pass that too

I tried a quick fix quickly - sorry accidentally pushed a commit to this branch. Meant a separate branch `t0yv0/empty_set_cross_test`

I can get the quick fix to pass RawConfig test but curiously NOT the RawPlan test. This makes me wonder what's going on there. Could this be because we'e linking in outdated code for planning the change compared to TF CLI? It's a bit unfortunate that in this test setup we cannot debug into TF CLI to see what is it doing .

t0yv0 commented:
I'd also wager my fix doesn't cut it for the attribute case which might need to actually return `null` not `[]`.

t0yv0 commented:
BTW the reason this may be less impactful than it seems is that few providers consult RawPlan and RawConfig on Create, they mostly consult the ResourceData. So issues of this sort remain unfixed.

VenelinMartinov commented:
I think the attribute case is completely unhandled but let's fix that seprately.

Isn't there the same issue in https://github.com/pulumi/pulumi-terraform-bridge/blob/26a293813b6274f73a7cdb36390961346b20877c/pkg/tfshim/sdk-v2/proposed_new.go#L99

t0yv0 commented:
Well that one looks fine? hcty2ctyWithType converts between two equivalent representations cty.Value and hcty.Value, converting null to null is reasonable, why do we have null passed into that in the first place would be the question?

Perhaps we need to fix at the level of MakeTerraformInputs? 

What is not yet tested that's really interesting to test, even more fundamentally important than RawConfig/RawPLan parity, is whether the two paths get the same answer to this:

```
tfResData.Get("f0")
tfResData.GetChange("f0")
```

VenelinMartinov commented:
I believe I found two other places where the state is populated with nulls instead of empty collections:

https://github.com/pulumi/pulumi-terraform-bridge/blob/012ac655189047bcd62bc4b7f25104d8f072da0d/pkg/tfshim/sdk-v2/provider2.go#L307

and https://github.com/pulumi/pulumi-terraform-bridge/blob/012ac655189047bcd62bc4b7f25104d8f072da0d/pkg/tfshim/sdk-v2/provider2.go#L334

The first one is fairly straight-forward to fix but the upgradeState stuff might be better fixed by calling a higher level function in the plugin-sdk -it looks to me like we are doing multiple passes back and forth between instanceState and cty values both in the bridge code and then in the plugin-sdk.

VenelinMartinov commented:
Ah, I think we are missing a null normalization after calling ...

In the plugin-sdk: https://github.com/hashicorp/terraform-plugin-sdk/blob/70fb6b9b15e8e5fc73f424e24084c28fedd1e013/helper/schema/grpc_provider.go#L1137-L1145

`StateValueFromInstanceState` is just a wrapper around `AttrsAsObjectValue`


But in `provider2.go`: https://github.com/pulumi/pulumi-terraform-bridge/blob/012ac655189047bcd62bc4b7f25104d8f072da0d/pkg/tfshim/sdk-v2/provider2.go#L358

we don't then normalize the nulls to empty collections.

VenelinMartinov commented:
Of interest here is this plugin-sdk function: https://github.com/pulumi/terraform-plugin-sdk/blob/70fb6b9b15e8e5fc73f424e24084c28fedd1e013/helper/schema/grpc_provider.go#L1502

Has some valuable notes there.

VenelinMartinov commented:
Ran downstream tests. Got quite a few failures in AWS, mostly related to security groups. Will investigate

FAILURE https://github.com/pulumi/pulumi-aws/pull/3955
FAILURE https://github.com/pulumi/pulumi-gcp/pull/2007
FAILURE https://github.com/pulumi/pulumi-hcloud/pull/509

VenelinMartinov commented:
I think I'll split the provider2 UpgradeState changes from the cty changes and ship them separately.

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -322,31 +323,70 @@ func (p *planResourceChangeImpl) upgradeState(
+	// https://github.com/opentofu/opentofu/blob/2ef3047ec6bb266e8d91c55519967212c1a0975d/internal/tofu/upgrade_resource_state.go#L52
+	if version > int64(res.SchemaVersion) {
+		return nil, fmt.Errorf(
+			"State version %d is greater than schema version %d for resource %s. "+
```
NIce

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -322,31 +323,70 @@ func (p *planResourceChangeImpl) upgradeState(
+		)
+	}
+
+	// Note upgrade is always called, even if the versions match
```
Thanks for checking this.

t0yv0 commented on pkg/tfshim/sdk-v2/cty_test.go:
```diff
@@ -821,6 +826,12 @@ func normalizeCtyValue(v cty.Value) cty.Value {
 			f.SetPrec(512)
 			return cty.NumberVal(f), nil
 		}
+		if v.Type().IsSetType() && v.IsNull() {
```
What would it take to not have these changes? This significantly weakens the tests as it says that the tests don't care anymore about the difference between null and empty sets.

iwahbe reviewed (commented):
This looks very promising. I would echo https://github.com/pulumi/pulumi-terraform-bridge/pull/1971/files#r1605064804 before we merge.

VenelinMartinov commented:
The downstream errors I've checked so far all look like issues with Optional + Computed lists.

`TestListOptionalComputedNoChange` should repro the issue.

t0yv0 commented:
I have a strong hunch this is because attributes vs blocks distinction is not handled, as indeed, it must!

Make it a block and it passes fine:

```golang
func TestOptionalComputedCollectionBlocksNoChange(t *testing.T) {
	skipUnlessLinux(t)
	config := tftypes.NewValue(tftypes.Object{
		AttributeTypes: map[string]tftypes.Type{
			"security_groups": tftypes.List{
				ElementType: tftypes.Object{
					AttributeTypes: map[string]tftypes.Type{
						"foo": tftypes.String,
					},
					OptionalAttributes: map[string]struct{}{
						"foo": {},
					},
				},
			},
		},
		OptionalAttributes: map[string]struct{}{
			"security_groups": {},
		},
	}, map[string]tftypes.Value{})

	for _, tc := range []struct {
		name     string
		maxItems int
		typ      schema.ValueType
	}{
		// {"list", 0, schema.TypeList},
		// {"set", 0, schema.TypeSet},
		{"list max items one", 1, schema.TypeList},
		// {"set max items one", 1, schema.TypeSet},
	} {
		t.Run(tc.name, func(t *testing.T) {
			runDiffCheck(t, diffTestCase{
				Resource: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"security_groups": {
							Type:     tc.typ,
							Optional: true,
							Computed: true,
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									"foo": &schema.Schema{
										Optional: true,
										Type:     schema.TypeString,
									},
								},
							},
							MaxItems: tc.maxItems,
						},
					},
					CreateContext: func(
						ctx context.Context, rd *schema.ResourceData, i interface{},
					) diag.Diagnostics {
						err := rd.Set("security_groups", []any{map[string]any{"foo": "sg1"}})
						require.NoError(t, err)
						rd.SetId("someid")
						return make(diag.Diagnostics, 0)
					},
				},
				Config1: config,
				Config2: config,
			})
		})
	}
}

```

The remedy here, I think, is that that nils should be replaced by empty collections if and only if we are dealing with a block. That should not happen for attributes.

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -847,3 +852,38 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+		if val.GetAttr(fieldName).IsNull() {
+			// normalize it
+			fieldType := val.Type().AttributeType(fieldName)
+			// TODO: Can we assume InternalValidate passes on the schema?
```
Yes, bridge enforces that at build time, fair to assume!

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -847,3 +852,38 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+			}
+		}
+	}
+	// TODO nested blocks?
```
Pretty important..

VenelinMartinov commented:
The `TestInputsNestedBlocksEmpty` test does not produce the correct TF program, which is likely the same issue as https://github.com/pulumi/pulumi-terraform-bridge/issues/2021

I have some other tests which fail to produce the correct programs in https://github.com/pulumi/pulumi-terraform-bridge/pull/2022.

I'll address that next in order to make sure we can sufficiently test any solution here.

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -833,8 +831,53 @@ func normalizeBlockCollections(val cty.Value, res *schema.Resource) cty.Value {
 }
+
+func normalizeIterable(blockVal cty.Value, blockRes *schema.Resource) []cty.Value {
+	if blockVal.IsNull() ||
```
there is the interesting question of unknowns. You may be assuming here that MakeTerraformInputs has taken care of unknowns by this point, but if so it's good to assert. Otherwise will just panic in  blockVal.AsValueSlice() unhepfully.

VenelinMartinov replied:
Good point, thanks!

VenelinMartinov commented:
The latest round of failures are somewhat curious: https://github.com/pulumi/pulumi-terraform-bridge/issues/2025

VenelinMartinov commented:
I've kicked off downstream tests from [7a5e480](https://github.com/pulumi/pulumi-terraform-bridge/pull/1971/commits/7a5e480f2241302c7da49e39ec7cf9b0be7293e2), hopefully everything passes.

The changes should be good to go but I'll wait for the downstream tests before announcing that.

VenelinMartinov commented:
Most of the downstream updates which weren't there before look like https://github.com/pulumi/pulumi-terraform-bridge/issues/2047

Empty maps refresh dirty now.

iwahbe commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -798,3 +803,85 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+}
+
+func normalizeIterable(blockVal cty.Value, blockRes *schema.Resource) []cty.Value {
+	if !blockVal.Type().IsListType() &&
+		!blockVal.Type().IsSetType() {
+		contract.Failf("normalizeIterable: Expected list or set type, got %v", blockVal.Type().GoString())
+	}
```
Nit: readability
```suggestion
	contract.Assertf(blockVal.Type().IsListType() || blockVal.Type().IsSetType(),
		"normalizeIterable: Expected list or set type, got %v", blockVal.Type().GoString())
```

iwahbe reviewed (commented):
I'm exited to get this in. If the downstream tests go well, does that mean we are ready to merge?

VenelinMartinov commented:
Yeah, the PR should be good to go if downstream tests pass.

VenelinMartinov commented:
Tested pulumi-aws after the ComputeID fixes - it flaked on one test but is otherwise fine: https://github.com/pulumi/pulumi-aws/actions/runs/9350339814/job/25736323726?pr=4012

I think this is good to go. I've seen to failures in downstream tests aside from the repos with ComputeID issues.

t0yv0 commented on pkg/tfshim/sdk-v2/provider.go:
```diff
@@ -62,6 +62,7 @@ func NewProvider(p *schema.Provider, opts ...providerOption) shim.Provider {
 		tf:   p,
 		opts: opts,
 	}
+
```
Nit: remove whitespace change to simplify the diff.

VenelinMartinov replied:
Mb, thanks for spotting.

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
 	return dd
 }
+
+func normalizeBlockCollections(val cty.Value, res *schema.Resource) cty.Value {
```
Can this tolerate val==null (I'm not sure if this ever arises, but it may during Create, worth checking). I don't think it needs to handle val==unknown, but can assert about it.

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+			// Only lists and sets can be blocks and pass InternalValidate
+			// Ignore other types.
+			if fieldType.IsListType() {
+				glog.V(10).Info("normalizing list %s, %s", fieldName, fieldType.ElementType())
```
Nit: normalizing is a bit of a generic term, may say something like "replacing a nil list with an empty list because the underlying TF property is a block"

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+			}
+			subBlockRes, ok := subBlockSchema.Elem.(*schema.Resource)
+			if !ok {
+				glog.V(5).Info("normalizeBlockCollections: Unexpected schema type %s", fieldName)
```
This branch doesn't sound exceptional. I think this can be an attribute with Elem=Schema or something with Elem=nil, fairly normal no? If the hunch is right perhaps we should be silent in this case, not log anything.

VenelinMartinov replied:
I believe it is exceptional since we are iterating over sch.BlockTypes and these should only contain the block types - this should mean Elem = Resource && Type == List || Set

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+				glog.V(5).Info("normalizeBlockCollections: Unexpected schema type %s", fieldName)
+				continue
+			}
+			glog.V(10).Info("normalizing block %s %s", fieldName, subVal.Type().GoString())
```
This sounds a bit verbose, maybe leave logs to locations that actually change something?

t0yv0 commented on pkg/tfshim/sdk-v2/provider2.go:
```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
+func normalizeIterable(blockVal cty.Value, blockRes *schema.Resource) []cty.Value {
+	contract.Assertf(blockVal.Type().IsListType() || blockVal.Type().IsSetType(),
+		"normalizeIterable: Expected list or set type, got %v", blockVal.Type().GoString())
+	if blockVal.IsNull() || !blockVal.IsKnown() {
```
Are we certain about the  `!blockVal.IsKnown()` case? This seems like a bigger deal. Tested this well?

VenelinMartinov replied:
I've opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2032 to follow up with an investigation of unknown blocks, which is inexpressible in TF.

It should not happen but if it does we can't do anything about it in that function, so I've opted to keep it as is for now.

t0yv0 commented:
I would like to understand test coverage here at the integration level.

Something changed with the cross-tests, perhaps for the better, but it's not visible in the diff. Can we get a summary? Perhaps double up on some of  these with targeted test cases?

fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1970
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1915
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1964
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1767
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1762

The most worrisome case to me is replacing unknowns with empty collections, basically testing this at an integration level, since I believe we have several subsystems fighting over it (MakeTerraformInputs, and now normalizeCollections). Can we find out how the bridged provider behavior is changing in that case, and importantly, why, is this what we want?

https://github.com/pulumi/pulumi-terraform-bridge/issues/1885 related here but might not be the end of the story.

VenelinMartinov commented:
>The most worrisome case to me is replacing unknowns with empty collections, basically testing this at an integration level, since I believe we have several subsystems fighting over it (MakeTerraformInputs, and now normalizeCollections). Can we find out how the bridged provider behavior is changing in that case, and importantly, why, is this what we want?

I've removed the code which modified unknowns - it was wrong and unused.

I've also added GRPC tests for handling unknowns both in PlanResourceChange and non-PlanResourceChange. I've verified these have not changed between master and this PR. Very interesting how this has change for PlanResourceChange but unrelated to this PR.

I also added targeted regression tests for the fixed issues.

Let me know if this makes sense.

t0yv0 commented on pkg/tfbridge/provider_test.go:
```diff
@@ -4536,7 +4536,138 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
+	p := &schemav2.Provider{
+		Schema: map[string]*schemav2.Schema{},
+		ResourcesMap: map[string]*schemav2.Resource{
+			"example_resource": {
```
Doesn't seem to cover blocks (Elem: &schemav2.Resource{..}), but only attributes. Blocks are indeed the interesting cases. Would it be possible to cover?

VenelinMartinov replied:
mb, thanks for noticing.

t0yv0 commented on pkg/tfbridge/provider_test.go:
```diff
@@ -4535,3 +4535,339 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
+		"response": {
+			"properties":{
+				"id":"",
+				"nestedListProps":["04da6b54-80e4-46f7-96ec-b56ff0331ba9"]
```
What's going on here

VenelinMartinov replied:
very curious but unrelated to the changes in this PR - I've verified the same behaviour on master

t0yv0 commented on pkg/tfbridge/provider_test.go:
```diff
@@ -4535,3 +4535,339 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
+				"id":"04da6b54-80e4-46f7-96ec-b56ff0331ba9",
+				"stringProp":null,
+				"setProps":null,
+				"listProps":"04da6b54-80e4-46f7-96ec-b56ff0331ba9",
```
The unknown is floating from `[unk]` to `unk`, strange. Let me debug this.

t0yv0 commented on pkg/tfbridge/provider_test.go:
```diff
@@ -4535,3 +4535,339 @@ func TestPlanResourceChangeStateUpgrade(t *testing.T) {
+				"setProps":null,
+				"listProps":null,
+				"maxItemsOneProp":null,
+				"nestedListProps":["04da6b54-80e4-46f7-96ec-b56ff0331ba9"]
```
Another case of floating.

t0yv0 commented:
Verified that `[unk]` to `unk` floating happens inside PlanResourceChange method so this is something programmed in TF it seems, we cannot easily avoid this without digging further. Let's not block on it but accept as is for now.

t0yv0 commented:
Let's check that `"listBlockProps":"04da6b54-80e4-46f7-96ec-b56ff0331ba9"` top-level unknown gives something reasonable and if so LGTM!

VenelinMartinov commented on pkg/tfbridge/provider_test.go:
```diff
@@ -5484,7 +5484,7 @@ func TestPlanResourceChangeUnknowns(t *testing.T) {
 				"setBlockProps":[],
 				"listBlockProps":[],
-				"nestedListBlockProps":[{"nestedProps":null}],
+				"nestedListBlockProps":[{"nestedProps":[]}],
```
This is the only change in behaviour to current master related to unknowns and it seems reasonable to me - nestedProps is a list type so should not be nil.

t0yv0 replied:
Yeah looks like MakeTerraformInputs substituted "04da6b54-80e4-46f7-96ec-b56ff0331ba9" into `"nestedListBlockProps":[{"nestedProps":null}],` and then the newly added pass corrected that to `[]`. This is fine.. we're already in heuristic land in this case. Good to have it covered!!

t0yv0 commented:
:shipit:
